lib/dal/src/attribute/value/dependent_value_graph.rs

```diff
@@ -14,77 +15,127 @@ use crate::{
+            // DependentValuesUpdate job may have been removed from the snapshot between when the
+            // DVU job was created and when we're processing things now. This could happen if there
+            // are other modifications to the snapshot before the DVU job starts executing, as the
+            // job always operates on the current state of the change set's snapshot, not the state
+            // at the time the job was created.
             if workspace_snapshot
-                .try_get_node_index_by_id(value_id)
+                .try_get_node_index_by_id(initial_id)
                 .await?
                 .is_none()
             {
-                debug!("Attribute Value {value_id} missing, skipping it in DependentValueGraph");
+                debug!(%initial_id, "missing node, skipping it in DependentValueGraph");
                 continue;
             }
 
-            let child_values = AttributeValue::all_object_children_to_leaves(ctx, value_id).await?;
-            values.push(value_id);
-            values.extend(child_values);
+            let node_weight = workspace_snapshot.get_node_weight_by_id(initial_id).await?;
+
+            match node_weight {
+                NodeWeight::AttributeValue(_) => {
+                    // We need to pre-process the initial values here to add any object
+                    // child values. We do this before the work loop instead of during so
+                    // that we only add the child values of the values we have been
+                    // explicitly asked to calculate the graph for, not any objects
+                    // encountered while walking the prototype chain
+                    let value: AttributeValueId = initial_id.into();
+                    let child_values =
+                        AttributeValue::all_object_children_to_leaves(ctx, value).await?;
+                    values.push(value);
+                    values.extend(child_values);
+                }
+                NodeWeight::Content(content_node_weight) => {
+                    match content_node_weight.content_address().into() {
+                        // If we are processing a secret, we don't want to add the secret itself.
+                        // We only want to add the direct dependent attribute values. We also need
+                        // to mark them as exceptions to the "independent values" collection.
+                        ContentAddressDiscriminants::Secret => {
+                            let direct_dependents =
+                                Secret::direct_dependent_attribute_values(ctx, initial_id.into())
+                                    .await
+                                    .map_err(Box::new)?;
+                            self.exceptions_to_independent_values
+                                .extend(direct_dependents.clone());
+                            values.extend(direct_dependents);
```

Comment:
Two subtle things here:

1. Unlike the above, we _do not_ push the initial value into the `values` if it is a `SecretId`.
2. We are adding the first set of values as _exceptions_ to the `independent_values()` method. I don't know how I feel about the shape or naming, but this feels like the best way (as of now) to ensure that the DVU job is not affected by these graph generation changes.

Reply:
This is outdated